lib/puppet/indirector/file_bucket_file/file.rb

```diff
@@ -189,19 +210,45 @@ def path_for(bucket_path, digest, subfile = nil)
       Puppet::FileSystem.pathname(subfile ? ::File.join(basedir, subfile) : basedir)
     end
 
-    # @param contents_file [Object] Opaque file path
-    # @param bucket_file [IO]
-    def verify_identical_file!(contents_file, bucket_file)
-      if bucket_file.size == Puppet::FileSystem.size(contents_file)
-        if bucket_file.stream() {|s| Puppet::FileSystem.compare_stream(contents_file, s) }
-          Puppet.info "FileBucket got a duplicate file #{bucket_file.checksum}"
-          return
+    # @param contents_file [Pathname] Opaque file path to intended backup
+    #   location
+    # @param bucket_file [Puppet::FileBucket::File] IO object representing
+    #   content to back up
+    # @return [Boolean] whether the data in contents_file is of the same size
+    #   and content as that in the bucket_file
+    # @api private
+    def verify_identical_file(contents_file, bucket_file)
+      (bucket_file.size == Puppet::FileSystem.size(contents_file)) &&
+        (bucket_file.stream() {|s| Puppet::FileSystem.compare_stream(contents_file, s) })
+    end
+
+    # @param contents_file [Pathname] Opaque file path to intended backup
+    #   location
+    # @param expected_checksum_data [String] expected value of checksum of type
+    #   checksum_type
+    # @param checksum_type [String] type of check sum of checksum_data, ie "md5"
+    # @return [Boolean] whether the checksum of the contents_file matches the
+    #   supplied checksum
+    # @api private
+    def contents_file_matches_checksum?(contents_file, expected_checksum_data, checksum_type)
+      contents_file_checksum_data = Puppet::Util::Checksums.method(:"#{checksum_type}_file").call(contents_file.to_path)
+      contents_file_checksum_data == expected_checksum_data
+    end
+
+    # @param contents_file [Pathname] Opaque file path to intended backup
+    #   location
+    # @param bucket_file [Puppet::FileBucket::File] IO object representing
+    #   content to back up
+    # @return [void]
+    # @api private
+    def copy_bucket_file_to_contents_file(contents_file, bucket_file)
+       Puppet::FileSystem.open(contents_file, 0440, 'wb') do |of|
+        # PUP-1044 writes all of the contents
+        bucket_file.stream() do |src|
+          FileUtils.copy_stream(src, of)
```

Comment:
I was initially worried about that possibility as well. I was just more worried about changing more behavior than just fixing the empty filebucket problem, and accidentally introducing some unrelated regression - this is the same behavior that existed before my PR: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/indirector/file_bucket_file/file.rb#L154-L158. That said, when I looked at it before, I concluded we do actually protect against multiple simultaneous writes because we obtain a lock on the corresponding paths file before beginning the contents write: https://github.com/MosesMendoza/puppet/blob/c72777242b8571813bce4a2c55a9f1e78b4a4fbf/lib/puppet/indirector/file_bucket_file/file.rb#L157. I think through that exclusive lock is the only path by which we'd be writing to the contents file. However, it was much clearer that this is the case (and thus probably safer) before I broke this out into a separate method to reduce duplication.

However, it occurs to me now - I think that ruby's `::File::LOCK_EX` (used by our `Puppet::FileSystem::FileImpl#exclusive_open` implementation) allows you to restrict access to one process, which was safe in rack/passenger days, but I think Puppet Server rubies live in different threads in the same process.. so is this broken there?